AK/AllOf.h

```diff
@@ -17,22 +18,16 @@ constexpr bool all_of(
     TEndIterator const& end,
     auto const& predicate)
 {
-    for (auto iter = begin; iter != end; ++iter) {
-        if (!predicate(*iter)) {
-            return false;
-        }
-    }
-    return true;
+    constexpr auto negated_predicate = [](auto const& pred) {
+        return [&](auto const& elem) { return !pred(elem); };
+    };
+    return find_if(begin, end, negated_predicate(predicate)) == end;
```

Comment:
Technically speaking, `TIterator::operator==(TEndIterator const&) const` is not in the requirements for `IteratorPairWith<TEndIterator>`, this currently works because our tests don't test not having an `operator==` (that's used to let the compiler derive `operator!=`).
Let's just work around that by:
```suggestion
    return !(find_if(begin, end, negated_predicate(predicate)) != end);
```

sidenote: does this `negated_predicate` have any impact? Ideally I'd expect to see it be inlined, but I haven't checked - have you?

Reply:
Thanks for the suggestion on the `operator==()`. I will address that.

On the performance note, I had assumed it would be the same, but it is not. It actually optimizes *better* with the `negated_predicate`. Wow! (note, these are using STL, not AK, but should be similar or identical)

Quick-bench: https://quick-bench.com/q/QVfhrCpR2vKeQskeo_MWmvTV5OI

Here is the assembly output too. Note that while it generates larger code, it ends up running faster in the micro-benchmark. That also holds when it is a non-power-of-2 value.

Original: https://serenityos.godbolt.org/z/5Tv7nGqvv
Passing predicate directly with capture: https://serenityos.godbolt.org/z/65T7PdWhe
Passing negated predicate: https://serenityos.godbolt.org/z/Kedn45vjf

Reply:
The speedup you see is specific to `std::find_if`, which seems to be manually unrolled (4 per iter) - see `bits/stl_algobase.h` (`__find_if(_RandomAccessIterator __first, _RandomAccessIterator __last, _Predicate __pred, random_access_iterator_tag)`).

Not including that unrolling behaviour produces the same code, which is fine.
but please don't bring random unexpected variables into a statistics fight :P